mantis-control-plane/mantis-control-plane-dynamodb/src/main/java/io/mantisrx/extensions/dynamodb/DynamoDBLeaderElector.java

```diff
@@ -0,0 +1,151 @@
+import com.amazonaws.services.dynamodbv2.LockItem;
+import io.mantisrx.server.core.BaseService;
+import io.mantisrx.server.core.ILeadershipManager;
+import io.mantisrx.server.core.json.DefaultObjectMapper;
+import io.mantisrx.server.core.master.MasterDescription;
+import io.mantisrx.shaded.com.fasterxml.jackson.core.JsonProcessingException;
+import io.mantisrx.shaded.com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DynamoDBLeaderElector extends BaseService {
+
+    private static final Logger logger = LoggerFactory.getLogger(DynamoDBLeaderElector.class);
+
+    private final ThreadFactory leaderThreadFactory = r -> {
+        Thread thread = new Thread(r);
+        thread.setName("dynamodb-leader-" + System.currentTimeMillis());
+        thread.setDaemon(true); // allow JVM to shutdown if monitor is still running
+        thread.setPriority(Thread.NORM_PRIORITY);
+        thread.setUncaughtExceptionHandler((t, e) -> logger.error("thread: {} failed with {}", t.getName(), e.getMessage(), e) );
+        return thread;
+    };
+    private final ScheduledExecutorService leaderElector =
+            Executors.newSingleThreadScheduledExecutor(leaderThreadFactory);
+    private final AtomicBoolean shouldLeaderElectorBeRunning = new AtomicBoolean(true);
+
+    private final AtomicBoolean isLeaderElectorRunning = new AtomicBoolean(false);
+
+    private final ObjectMapper jsonMapper = DefaultObjectMapper.getInstance();
+    private final ILeadershipManager leadershipManager;
+    private final AmazonDynamoDBLockClient lockClient;
+    private final String partitionKey;
+
+    @Nullable
+    private LockItem leaderLock = null;
+
+    public DynamoDBLeaderElector(ILeadershipManager leadershipManager) {
+        this(leadershipManager,
+            DynamoDBClientSingleton.getLockClient(),
+            DynamoDBClientSingleton.getPartitionKey());
+
+    }
+    public DynamoDBLeaderElector(
+            ILeadershipManager leadershipManager,
+            AmazonDynamoDBLockClient lockClient,
+            String key) {
+        this.leadershipManager = leadershipManager;
+        this.lockClient = lockClient;
+        this.partitionKey = key;
+    }
+
+    @Override
+    @SuppressWarnings("FutureReturnValueIgnored")
+    public void start() {
+       if(!isLeaderElectorRunning() && shouldLeaderElectorBeRunning.get()) {
+           logger.info("starting leader elector");
+           leaderElector.submit(this::tryToBecomeLeader);
+       }
+    }
+
+    @Override
+    public void shutdown() {
+        logger.info("shutting down");
```

Comment:
I debated this shutdown order a few times. This is difficult to test, but I assume since leaders kill themselves the procedure can be best effort.